tt_metal/llrt/index_map.hpp

```diff
@@ -0,0 +1,121 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <vector>
+#include <optional>
+
+namespace tt {
+
+/**
+ * IndexMap is a container, which mimics std::map with an assumption that all keys are indexes.
+ * Provides vector-like performance, consumes the size proportional to the maximum used index.
+ * Implemented as an auto-resizing vector of optionals.
+ */
+template <typename T>
+class IndexMap {
+public:
```

Comment:
`boost::flat_map` and `boost::unordered_flat_map` is a different kind of container. I want to the `operator[]` just to index into a vector without any overhead.

Reply:
It would be interesting to see a benchmark (quick-bench.com or similar ) for how this implementation compares to the boost ones for your use case.

Reply:
Benchmarking unordered map is quite complicated, the perf depends on usage-patterns, hash-collisions, etc.
We know that the vector is always faster, measuring how much faster is quite involved. IndexMap is basically a vector of optionals, not much more to it. Based on the other reviews, it seems like we might be able to avoid it and use a simple vector.